banyand/liaison/grpc/server.go

```diff
@@ -323,6 +341,186 @@ func (s *server) GracefulStop() {
+func (tr *TLSReloader) Start() error {
+	tr.log.Info().Str("certFile", tr.certFile).Str("keyFile", tr.keyFile).Msg("Starting TLS file monitoring for gRPC")
+	err := tr.watcher.Add(tr.certFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to watch cert file: %s", tr.certFile)
+	}
+	err = tr.watcher.Add(tr.keyFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to watch key file: %s", tr.keyFile)
+	}
+	go tr.watchFiles()
+	return nil
+}
+
+// watchFiles monitors the certificate and key files for changes and reloads credentials.
+func (tr *TLSReloader) watchFiles() {
+	tr.log.Info().Msg("TLS file watcher loop started for gRPC")
+	for {
+		select {
+		case event, ok := <-tr.watcher.Events:
+			if !ok {
+				tr.log.Warn().Msg("Watcher events channel closed unexpectedly")
+				return
+			}
+			if event.Op&fsnotify.Remove == fsnotify.Remove {
+				if event.Name == tr.certFile {
+					if err := tr.watcher.Add(tr.certFile); err != nil {
+						tr.log.Error().Err(err).Str("file", tr.certFile).Msg("Failed to re-add cert file to watcher")
+					}
+				} else if event.Name == tr.keyFile {
+					if err := tr.watcher.Add(tr.keyFile); err != nil {
+						tr.log.Error().Err(err).Str("file", tr.keyFile).Msg("Failed to re-add key file to watcher")
+					}
+				}
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write ||
+				event.Op&fsnotify.Rename == fsnotify.Rename ||
+				event.Op&fsnotify.Create == fsnotify.Create {
+				tr.log.Info().Str("file", event.Name).Msg("Detected uploaded/changed certificate")
+				if err := tr.reloadCertificate(); err != nil {
+					tr.log.Error().Err(err).Str("file", event.Name).Msg("Failed to reload TLS certificate")
+				} else {
+					tr.log.Info().Str("file", event.Name).Msg("Successfully updated TLS certificate")
+				}
+			}
+		case err, ok := <-tr.watcher.Errors:
+			if !ok {
+				tr.log.Warn().Msg("Watcher errors channel closed unexpectedly")
+				return
+			}
+			tr.log.Error().Err(err).Msg("Error in file watcher")
+		case <-tr.stopCh:
+			tr.log.Info().Msg("Stopping TLS file watcher for gRPC")
+			return
+		}
+	}
+}
+
+// reloadCertificate reloads the TLS certificate from the certificate and key files.
+func (tr *TLSReloader) reloadCertificate() error {
+	newCert, err := tls.LoadX509KeyPair(tr.certFile, tr.keyFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to reload TLS certificate")
+	}
+	tr.mu.Lock()
+	defer tr.mu.Unlock()
+	tr.cert = &newCert
+	return nil
+}
+
+// GetCertificate returns the current TLS certificate.
+func (tr *TLSReloader) GetCertificate() *tls.Certificate {
+	tr.mu.RLock()
+	defer tr.mu.RUnlock()
+	return tr.cert
+}
+
+// Stop gracefully stops the TLS reloader.
+func (tr *TLSReloader) Stop() {
+	close(tr.stopCh)
+	if err := tr.watcher.Close(); err != nil {
+		tr.log.Error().Err(err).Msg("Failed to close fsnotify watcher")
+	}
+}
+
+// DynamicTLSCredentials implements credentials.TransportCredentials for dynamic reloading.
+type DynamicTLSCredentials struct {
```

Comment:
I prefer using `tls.NewTLS` to initialize "TransportCredentials" instead of a custom one. You can dynamically load certificates using "GetCertificate" in `tls.Config`.